### DISCUSSION TITLE
Loading WinUI Types in PowerShell

### DISCUSSION THREAD

**Discussion by nkasco** [microsoft/microsoft-ui-xaml#8206] [Q&A]
Sysadmins often create simpler UI based tools entirely within PowerShell, today with WinForms or WPF, but would love to be able to create something a bit more modern using WinUI 3. The reason why it's more desirable to do entirely within PowerShell rather than a C# app that then loads PowerShell is largely that Sysadmins are simply more comfortable staying within that language. 

It's my understanding that I'd need to load the C# projection assemblies to have access to the metadata based types within PowerShell 7.

After some digging, I found this deep dive that helps lay out some of the dependencies and it appears to state there is a projection assembly already created?
https://blogs.windows.com/windowsdeveloper/2020/07/07/a-deep-dive-into-winui-3-in-desktop-apps/#:~:text=It%E2%80%99s%20the%20Microsoft.WinUI.dll%20library.%20This%20library%20is%20included%20in%20the%20WinUI%203%20NuGet%20Package

With that in mind, I attempted the following:
`add-type -Path "path\to\Microsoft.Windows.SDK.NET.dll"`
`add-type -Path "path\to\WinRT.Runtime.dll"`
`add-type -Path "path\to\Microsoft.WinUI.dll"`

Which I obtained by creating a new solution in Visual Studio, building, and then extracting those dlls. All of those assemblies successfully load into my session without issue.

Once I load those, I seem to have access to the types: 
![image](https://user-images.githubusercontent.com/1930697/219872211-a1e8aa29-780d-474d-abc7-86f2a7ca0620.png)

However, when I try to create a new object I get the following error:
`MethodInvocationException: Exception calling ".ctor" with "0" argument(s): "The type initializer for '_IWindowFactory' threw an exception."`

I've tried to create objects using both the XamlReader Load method and New() straight from the type:
![image](https://user-images.githubusercontent.com/1930697/219872652-cc7343b5-125f-4a94-95e6-508e14b6dcb3.png)

Same thing even if I try to create a standalone textbox from Microsoft.UI.Xaml.Controls.Textbox, etc. I think I'm just missing a few dependencies, but I'm struggling to figure out where to go from here.

Can anyone on this side of the fence assist with a simple Hello World implementation to load this into PowerShell?

**Comment by DarranRowe** [comment]
There are two things that normally happen before you can start loading Xaml content.
1) Bootstrap the Windows App SDK. Even if you reference the .NET assemblies, that doesn't automatically load the Windows App SDK package.
2) Create a new WinRT type that derives from Microsoft.UI.Xaml.Application, this class must also implement the Microsoft.UI.Xaml.Markup.IXamlMetadataProvider interface. This must then provide information from Microsoft.UI.Xaml.XamlTypeInfo.XamlControlsXamlMetadataProvider at the very least.

The thread which displays content is also expected to have a Microsoft.UI.Dispatching.DispatcherQueue on it. I know that one is made during the initialisation, but I'm not sure exactly when.

But yes, that runtime class derived from Application is a real sticking point.

**Reply by nkasco** [reply]
Helpful info, any chance you might be able to provide a sample script?

**Reply by HotCakeX** [reply]
@nkasco Hi, have you found any answers to this? I want to do the same thing but can't find any info.

**Reply by nkasco** [reply]
@HotCakeX I went as far down the rabbit hole as I could for weeks and got essentially nowhere meaningful with it. 

I ended up starting a C# project to better serve my desire to launch scripts with style (and learn more about WinUI 3 in the process), it's available here: [https://github.com/nkasco/IT-Admin-Toolkit-WinUI](https://github.com/nkasco/IT-Admin-Toolkit-WinUI)

**Comment by driver1998** [comment]
A C# demo with no XAML may get you started?
https://github.com/driver1998/WinUI-Without-Xaml-CSharp
